## <b> Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!

## <b>This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data. </b>

## <b> Explore and analyze the data to discover important factors that govern the bookings. </b>

In [108]:
# Import Python libraries.
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [109]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [110]:
#Linked our Googlecolab notebook to Google Drive.
from google.colab import drive
drive.mount('//content//drive')

Drive already mounted at //content//drive; to attempt to forcibly remount, call drive.mount("//content//drive", force_remount=True).


In [111]:
#Data from Hotel Booking.csv file from Google Drive has been assigned to data frame named df_hotel. 
df_hotel=pd.read_csv("//content//drive//MyDrive//AlmaBetter//Hotel Bookings.csv")

df1=df_hotel.copy()

# ***Understand dataset input***

* Find out the total columns and rows of dataset
* Find the data type of each column.
* Find the continuous and categorical data
* Find individual distribution for some of the columns
* Also check the correlation between dependent columns

In [112]:
#Just take a look at the df_hotel file data to find out the total columns and rows of dataset.
df_hotel

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
0                          0       2  ...    No Deposit    NaN     NaN   
1                          0       2  ...    No Deposit    NaN     NaN   
2                          1       1  ...    No Deposit    NaN     NaN   
3                          1       1  ...    No Deposit  304.0     NaN   
4                          2       2  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
0                         0     Transient    0.00   
1                         0     Transient    0.00   
2                         0     Transient   75.00   
3                         0     Transient   75.00   
4                         0     Transient   98.00   
...                     ...           ...     ...   
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3             

In [113]:
#Get the basic info of each columns with not null rows and datatype which will give idea about the data type of each column.
df_hotel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            


# Now just take a look at the unique values of each column in the df_hotel file.


In [114]:
print(df_hotel['hotel'].unique())

['Resort Hotel' 'City Hotel']


There are two types of hotels
* Resort Hotel 
* City hotel

In [115]:
print(df_hotel['is_canceled'].unique())

[0 1]


This column predicts
* Booking cancelled = 1  
* Booking not cancelled = 0

In [116]:
print(df_hotel['lead_time'].unique())

[342 737   7  13  14   0   9  85  75  23  35  68  18  37  12  72 127  78
  48  60  77  99 118  95  96  69  45  40  15  36  43  70  16 107  47 113
  90  50  93  76   3   1  10   5  17  51  71  63  62 101   2  81 368 364
 324  79  21 109 102   4  98  92  26  73 115  86  52  29  30  33  32   8
 100  44  80  97  64  39  34  27  82  94 110 111  84  66 104  28 258 112
  65  67  55  88  54 292  83 105 280 394  24 103 366 249  22  91  11 108
 106  31  87  41 304 117  59  53  58 116  42 321  38  56  49 317   6  57
  19  25 315 123  46  89  61 312 299 130  74 298 119  20 286 136 129 124
 327 131 460 140 114 139 122 137 126 120 128 135 150 143 151 132 125 157
 147 138 156 164 346 159 160 161 333 381 149 154 297 163 314 155 323 340
 356 142 328 144 336 248 302 175 344 382 146 170 166 338 167 310 148 165
 172 171 145 121 178 305 173 152 354 347 158 185 349 183 352 177 200 192
 361 207 174 330 134 350 334 283 153 197 133 241 193 235 194 261 260 216
 169 209 238 215 141 189 187 223 284 214 202 211 16

* The column **Lead Time** predicts the time difference between guest books their room and the time they are scheduled to arrive at the hotel.

In [117]:
print(df_hotel['arrival_date_year'].unique())

[2015 2016 2017]


* **arrival_date_year** is Year of arrival of the guest.

In [118]:
print(df_hotel['arrival_date_month'].unique())

['July' 'August' 'September' 'October' 'November' 'December' 'January'
 'February' 'March' 'April' 'May' 'June']


* **arrival_date_month** is Months of arrival of the guest.

In [119]:
print(df_hotel['arrival_date_week_number'].unique())

[27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50
 51 52 53  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21
 22 23 24 25 26]


* **arrival_date_week_number** is weak of arrival of the guest.

In [120]:
print(df_hotel['meal'].unique())

['BB' 'FB' 'HB' 'SC' 'Undefined']


**Meal type**.

* **BB** : Bed and Breakfast
* **HB** : Half Board (Breakfast and Dinner normally)

* **FB** : Full Board (Beakfast, Lunch and Dinner)
* **SC** : Self-catering
* **Undefined** : Rooms only packages without meals.



In [121]:
print(df_hotel['market_segment'].unique())

['Direct' 'Corporate' 'Online TA' 'Offline TA/TO' 'Complementary' 'Groups'
 'Undefined' 'Aviation']


**Market segment** 
This provides source of information through which customer booked
* **TA** - "Travel Agent"
* **TO** - "Tour operators"
* **Direct** -"Direct booking

In [122]:
print(df_hotel['distribution_channel'].unique())


['Direct' 'Corporate' 'TA/TO' 'Undefined' 'GDS']


**Distribution Channel** 
* We provides source of information through which customer booked
* **TA/TO** - "Travel Agent"/"Tour operators"
* **Direct** -"Direct booking"
* **Corporate**- "Corporate booking"

In [123]:
print(df_hotel['is_repeated_guest'].unique())

[0 1]


**Guests are new or old.**
* **1**----> New guest
* **0**----> Old guest

In [124]:
print(df_hotel['reserved_room_type'].unique())

['C' 'A' 'D' 'E' 'G' 'F' 'H' 'L' 'P' 'B']


In [125]:
print(df_hotel['assigned_room_type'].unique())

['C' 'A' 'D' 'E' 'G' 'F' 'I' 'B' 'H' 'P' 'L' 'K']


In [126]:
print(df_hotel['booking_changes'].unique())

[ 3  4  0  1  2  5 17  6  8  7 10 16  9 13 12 20 14 15 11 21 18]


In [127]:
print(df_hotel['deposit_type'].unique())

['No Deposit' 'Refundable' 'Non Refund']


In [128]:
print(df_hotel['agent'].unique())

[ nan 304. 240. 303.  15. 241.   8. 250. 115.   5. 175. 134. 156. 243.
 242.   3. 105.  40. 147. 306. 184.  96.   2. 127.  95. 146.   9. 177.
   6. 143. 244. 149. 167. 300. 171. 305.  67. 196. 152. 142. 261. 104.
  36.  26.  29. 258. 110.  71. 181.  88. 251. 275.  69. 248. 208. 256.
 314. 126. 281. 273. 253. 185. 330. 334. 328. 326. 321. 324. 313.  38.
 155.  68. 335. 308. 332.  94. 348. 310. 339. 375.  66. 327. 387. 298.
  91. 245. 385. 257. 393. 168. 405. 249. 315.  75. 128. 307.  11. 436.
   1. 201. 183. 223. 368. 336. 291. 464. 411. 481.  10. 154. 468. 410.
 390. 440. 495. 492. 493. 434.  57. 531. 420. 483. 526. 472. 429.  16.
 446.  34.  78. 139. 252. 270.  47. 114. 301. 193. 182. 135. 350. 195.
 352. 355. 159. 363. 384. 360. 331. 367.  64. 406. 163. 414. 333. 427.
 431. 430. 426. 438. 433. 418. 441. 282. 432.  72. 450. 180. 454. 455.
  59. 451. 254. 358. 469. 165. 467. 510. 337. 476. 502. 527. 479. 508.
 535. 302. 497. 187.  13.   7.  27.  14.  22.  17.  28.  42.  20.  19.
  45. 

In [129]:
print(df_hotel['company'].unique())

[ nan 110. 113. 270. 178. 240. 154. 144. 307. 268.  59. 204. 312. 318.
  94. 174. 274. 195. 223. 317. 281. 118.  53. 286.  12.  47. 324. 342.
 373. 371. 383.  86.  82. 218.  88.  31. 397. 392. 405. 331. 367.  20.
  83. 416.  51. 395. 102.  34.  84. 360. 394. 457. 382. 461. 478. 386.
 112. 486. 421.   9. 308. 135. 224. 504. 269. 356. 498. 390. 513. 203.
 263. 477. 521. 169. 515. 445. 337. 251. 428. 292. 388. 130. 250. 355.
 254. 543. 531. 528.  62. 120.  42.  81. 116. 530. 103.  39.  16.  92.
  61. 501. 165. 291. 290.  43. 325. 192. 108. 200. 465. 287. 297. 490.
 482. 207. 282. 437. 225. 329. 272.  28.  77. 338.  72. 246. 319. 146.
 159. 380. 323. 511. 407. 278.  80. 403. 399.  14. 137. 343. 346. 347.
 349. 289. 351. 353.  54.  99. 358. 361. 362. 366. 372. 365. 277. 109.
 377. 379.  22. 378. 330. 364. 401. 232. 255. 384. 167. 212. 514. 391.
 400. 376. 402. 396. 302. 398.   6. 370. 369. 409. 168. 104. 408. 413.
 148.  10. 333. 419. 415. 424. 425. 423. 422. 435. 439. 442. 448. 443.
 454. 

In [130]:
print(df_hotel['days_in_waiting_list'].unique())

[  0  50  47  65 122  75 101 150 125  14  60  34 100  22 121  61  39   5
   1   8 107  43  52   2  11 142 116  13  44  97  83   4 113  18  20 185
  93 109   6  37 105 154  64  99  38  48  33  77  21  80  59  40  58  89
  53  49  69  87  91  57 111  79  98  85  63  15   3  41 224  31  56 187
 176  71  55  96 236 259 207 215 160 120  30  32  27  62  24 108 147 379
  70  35 178 330 223 174 162 391  68 193  10  76  16  28   9 165  17  25
  46   7  84 175 183  23 117  12  54  26  73  45  19  42  72  81  92  74
 167  36]


In [131]:
print(df_hotel['customer_type'].unique())

['Transient' 'Contract' 'Transient-Party' 'Group']


**This column predicts**
* **Transient**
* **Transient-Party**
* **Group**
* **Contract**

In [132]:
print(df_hotel['adr'].unique())

[  0.    75.    98.   ... 266.75 209.25 157.71]


In [133]:
print(df_hotel['required_car_parking_spaces'].unique())

[0 1 2 8 3]


In [134]:
print(df_hotel['total_of_special_requests'].unique())

[0 1 3 2 4 5]


In [135]:
print(df_hotel['reservation_status'].unique())

['Check-Out' 'Canceled' 'No-Show']


**This column predicts**
* **Check-Out**
* **Canceled**
* **No-Show**

# ***Cleaning the data***

* Identify duplicated rows and remove the same.
* Calculate percentage values of null vallues of each columns.
* Combine the null_value and null_value_percentage series in the dataframe using concat method.
* Replace NaN values with 0 for heading Agent,Company
* Replace NaN values with their mean values for heading children
* Modify datatype from float to int64 for heading Agent,Company,Children
* Replace NaN values with 'others' for heading Country.


In [136]:
#Identify duplicated rows
df_hotel[df_hotel.duplicated()]

hotel  is_canceled  lead_time  arrival_date_year  \
5       Resort Hotel            0         14               2015   
22      Resort Hotel            0         72               2015   
43      Resort Hotel            0         70               2015   
138     Resort Hotel            1          5               2015   
200     Resort Hotel            0          0               2015   
...              ...          ...        ...                ...   
119349    City Hotel            0        186               2017   
119352    City Hotel            0         63               2017   
119353    City Hotel            0         63               2017   
119354    City Hotel            0         63               2017   
119373    City Hotel            0        175               2017   

       arrival_date_month  arrival_date_week_number  \
5                    July                        27   
22                   July                        27   
43                   July                        27   
138                  July                        28   
200                  July                        28   
...                   ...                       ...   
119349             August                        35   
119352             August                        35   
119353             August                        35   
119354             August                        35   
119373             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
5                               1                        0   
22                              1                        2   
43                              2                        2   
138                             5                        1   
200                             7                        0   
...                           ...                      ...   
119349                         31                        0   
119352                         31                        0   
119353                         31                        0   
119354                         31                        0   
119373                         31                        1   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
5                          2       2  ...    No Deposit  240.0     NaN   
22                         4       2  ...    No Deposit  250.0     NaN   
43                         3       2  ...    No Deposit  250.0     NaN   
138                        0       2  ...    No Deposit  240.0     NaN   
200                        1       1  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119349                     3       2  ...    No Deposit    9.0     NaN   
119352                     3       3  ...    No Deposit    9.0     NaN   
119353                     3       3  ...    No Deposit    9.0     NaN   
119354                     3       3  ...    No Deposit    9.0     NaN   
119373                     3       1  ...    No Deposit   42.0     NaN   

       days_in_waiting_list    customer_type     adr  \
5                         0        Transient   98.00   
22                        0        Transient   84.67   
43                        0        Transient  137.00   
138                       0        Transient   97.00   
200                       0        Transient  109.80   
...                     ...              ...     ...   
119349                    0        Transient  126.00   
119352                    0  Transient-Party  195.33   
119353                    0  Transient-Party  195.33   
119354                    0  Transient-Party  195.33   
119373                    0        Transient   82.35   

        required_car_parking_spaces  total_of_special_requests  \
5                                 0                          1   
22                                0                          1   
43                                0         

In [137]:
#Remove duplicate rows.
df_hotel = df_hotel.drop_duplicates()
df_hotel.shape


(87396, 32)

In [138]:
#Find null values in each column and sum them.
null_value = df_hotel.isnull().sum()
print(null_value)

hotel                                 0
is_canceled                           0
lead_time                             0
arrival_date_year                     0
arrival_date_month                    0
arrival_date_week_number              0
arrival_date_day_of_month             0
stays_in_weekend_nights               0
stays_in_week_nights                  0
adults                                0
children                              4
babies                                0
meal                                  0
country                             452
market_segment                        0
distribution_channel                  0
is_repeated_guest                     0
previous_cancellations                0
previous_bookings_not_canceled        0
reserved_room_type                    0
assigned_room_type                    0
booking_changes                       0
deposit_type                          0
agent                             12193
company                           82137


In [139]:
#Calculate percentage values of null vallues of each columns.
null_value_percentage =(df_hotel.isnull().sum()/len(df_hotel))*100  #sum
print(null_value_percentage)

hotel                              0.000000
is_canceled                        0.000000
lead_time                          0.000000
arrival_date_year                  0.000000
arrival_date_month                 0.000000
arrival_date_week_number           0.000000
arrival_date_day_of_month          0.000000
stays_in_weekend_nights            0.000000
stays_in_week_nights               0.000000
adults                             0.000000
children                           0.004577
babies                             0.000000
meal                               0.000000
country                            0.517186
market_segment                     0.000000
distribution_channel               0.000000
is_repeated_guest                  0.000000
previous_cancellations             0.000000
previous_bookings_not_canceled     0.000000
reserved_room_type                 0.000000
assigned_room_type                 0.000000
booking_changes                    0.000000
deposit_type                    

In [140]:
#Combine the null_value and null_value_percentage series in the dataframe using concat method.
concatenate_null =pd.concat([null_value,null_value_percentage],axis=1,keys=['null_value','null_value_percentage'])
print(concatenate_null)

                                null_value  null_value_percentage
hotel                                    0               0.000000
is_canceled                              0               0.000000
lead_time                                0               0.000000
arrival_date_year                        0               0.000000
arrival_date_month                       0               0.000000
arrival_date_week_number                 0               0.000000
arrival_date_day_of_month                0               0.000000
stays_in_weekend_nights                  0               0.000000
stays_in_week_nights                     0               0.000000
adults                                   0               0.000000
children                                 4               0.004577
babies                                   0               0.000000
meal                                     0               0.000000
country                                452               0.517186
market_seg

* Here we observe that out of **32** columns, **4** columns have null values. These are the **children, country, agent, and company**.

In [141]:
#Check how many null values are there in children column.
df_hotel['children'].notnull().value_counts()

True     87392
False        4
Name: children, dtype: int64

In [142]:
#NaN values in children column which will be replaced by their mean value .
df_hotel["children"]=df_hotel["children"].fillna(df_hotel["children"].mean())


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [143]:
#Recheck 
df_hotel['children'].notnull().value_counts()
df_hotel['children']=df_hotel['children'].astype(int)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [144]:
#Check how many null values are there in country column.
df_hotel['country'].notnull().value_counts()

True     86944
False      452
Name: country, dtype: int64

In [145]:
#NaN values in country column which will be replaced by 'Other'
df_hotel['country'] = df_hotel['country'].fillna('other',axis=0)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [146]:
#Recheck
df_hotel['country'].notnull().value_counts()


True    87396
Name: country, dtype: int64

In [147]:
#Check how many null values are there in agent column.
df_hotel['agent'].notnull().value_counts()

True     75203
False    12193
Name: agent, dtype: int64

In [148]:
#NaN values in agent column will be replaced by 0. 
df_hotel['agent'] = df_hotel['agent'].fillna(0,axis=0)


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:
#Recheck
df_hotel['agent'].notnull().value_counts()
df_hotel['agent']=df_hotel['agent'].astype(int)

In [ ]:
#Check how many null values are there in company column.
df_hotel['company'].notnull().value_counts()

In [ ]:
#NaN values in agent column will be replaced by 0. 
df_hotel['company'] = df_hotel['company'].fillna(0,axis=0)

In [ ]:
#Recheck
df_hotel['company'].notnull().value_counts()
df_hotel['company']=df_hotel['company'].astype(int)

# **Data analysis and visualization** 

# Hotel type analysis
* Q1.Find type of hotel
* Q2.Find count of booking as per type of hotel

In [ ]:
#Q1.Find type of hotel
df_hotel['hotel'].unique()

In [ ]:
#Q2.Find count of booking as per type of hotel
hotelbook = df_hotel['hotel'].value_counts().reset_index()
hotelbook.columns=['Hotel type','Number of booking']
hotelbook

In [ ]:
#Graph
plt.pie(hotelbook["Number of booking"],autopct="%1.1f%%",radius=1.5,explode=[0.08]*2)
plt.legend(labels=hotelbook['Hotel type'].value_counts().index.tolist(),loc=(1.1,0.65))
plt.show()


* **Booking of city hotel[61.1%] is more preferable as compared to resort hotel[38.9%].**

# Find peak business season of hotel booking
* Q1. Display number of booking months wise
* Q2. Diaply no. of booking as per month and year.

In [ ]:
#Q1. Display number of booking months wise
month_wise_data= df_hotel['arrival_date_month'].value_counts()
plt.figure(figsize=(15,5))
month_wise_data.plot(kind = "bar",color='r',grid=True)
plt.title('Display Monthwise data')
plt.xlabel('Months')
plt.ylabel('Number of Guests');

In [ ]:
# Q2. Diaply no. of booking as per month and year.
dfg = df_hotel.groupby(['hotel','arrival_date_month', 'arrival_date_year'])['is_canceled'].sum()
dfg.plot(kind='bar', title='Booking Peak season', ylabel='No of bookings',
         xlabel='Month & Year',  figsize=(20, 10))

* **As first barplot shows, August month is a peak season for hotel business.**
* **As second barplot shows, City hotel has done maximum bookings in month of May,2017.**

* **Hotel higher booking cancellation rate**


In [ ]:
#we find canceled counts
cancellation_rate = df_hotel['is_canceled'].value_counts().reset_index()
#Rename columns
cancellation_rate.columns=['cancellation_graph','User']
#plot graph
sns.barplot(x=cancellation_rate['cancellation_graph'],y=cancellation_rate['User'],palette="tab10")
plt.title("Cancellation Graph")
plt.ylabel("number of guest")
plt.xlabel("Booking not canceled                  Booking canceled")
plt.show()


# ADR(Average Daily Rate)
* Q1. Find total no. of guests on basis of adults,children and babies.
* Q2. Find ADR per person.
* Q3. Find total no. of stay on basis of week nights and weekend nights.
* Q4. Find hotel revenue


In [ ]:
#Q1. Find total no. of guests on basis of adults,children and babies.

ADR_For_NonCanceledBooking = df_hotel[df_hotel['is_canceled'] !=1]  
ADR_For_NonCanceledBooking['Kids'] = ADR_For_NonCanceledBooking.children + ADR_For_NonCanceledBooking.babies
ADR_For_NonCanceledBooking['Total_Guests'] = ADR_For_NonCanceledBooking.Kids + ADR_For_NonCanceledBooking.adults

#Q2. Find ADR per person.
ADR_For_NonCanceledBooking['ADR_pp'] = (ADR_For_NonCanceledBooking.adr/ADR_For_NonCanceledBooking.Total_Guests)

#Q3. Find total no. of stay on basis of week nights and weekend nights.
ADR_For_NonCanceledBooking['TotalDays'] = ADR_For_NonCanceledBooking.stays_in_weekend_nights + ADR_For_NonCanceledBooking.stays_in_week_nights

#Q4. Find hotel revenue
ADR_For_NonCanceledBooking['HotelRevenue'] = (ADR_For_NonCanceledBooking.ADR_pp*ADR_For_NonCanceledBooking.TotalDays)

print (ADR_For_NonCanceledBooking)
sns.lineplot(data = ADR_For_NonCanceledBooking, x = 'arrival_date_month', y = 'HotelRevenue', hue = 'hotel')
plt.show()

* **Resort hotel is getting more revenue in the month of August.**

# Type of room
* Q1.How many types of rooms are provided through Hotels? 
* Q2.Which is most preffered room by the guest?


In [ ]:
#Q1.How many types of rooms are provided through Hotels? 
provided_room_by_hotel = df_hotel['reserved_room_type'].unique()
provided_room_by_hotel


In [ ]:
#Q2.Which is most preffered room by the guest?
most_prefer_room = df_hotel['reserved_room_type'].value_counts().reset_index()
most_prefer_room.columns=['room_type','number_of_prefrence']

sns.barplot(x=most_prefer_room["room_type"],y=most_prefer_room["number_of_prefrence"],palette="Set2")
plt.title("Bar chart of room wise prefrence")
plt.ylabel("Number of Guest ")
plt.xlabel("Room Type")
plt.show()

* **"A" type room is more preferred by guests.**

#Meal Consumption Analysis
* Q1. Which type of meal offer by hotel?
* Q2. Which is the most preferred meal consumption by guests?

In [ ]:
#Q1. Which type of meal offer by hotel?
df_hotel['meal'].unique()

In [ ]:
#Q2. Which is the most preferred meal consumption by guests?
meal=df_hotel['meal'].value_counts().reset_index()
meal.columns=['Meal_type','number_of_preference']
meal

In [ ]:
#Graph 
sns.barplot(x=meal["Meal_type"],y=meal["number_of_preference"],palette="tab10")
plt.title('Graph of Meal Consumption')
plt.ylabel('Number of Meal Consumption User')
plt.xlabel("Meal Type")
plt.show()


* **"BB" meal type is mostly preferred by guests.**

## **Waiting time Analysis**
* Q1. Analyse in which type of hotel there is more waiting time
* Q2. Analyse the average of type of hotel there is more waiting time

In [ ]:
#Q1. Analyse in which type of hotel there is more waiting time
max_waiting_time=df_hotel.groupby("hotel").max()["days_in_waiting_list"].reset_index()
max_waiting_time.columns=["Hotel","Max_Waiting_Time"]
print (max_waiting_time)

plt.figure(figsize=(10,5))
plt.subplot(1,2,2)
sns.barplot(x=max_waiting_time["Hotel"],y=max_waiting_time["Max_Waiting_Time"],palette="Set2")
plt.title("Hotel type vs Maximum Waiting Time")
plt.ylabel("Maximum Waiting Time")
plt.xlabel("Hotel type")
plt.show()

In [ ]:
#Q2. Analyse the average of type of hotel there is more waiting time
avg_waiting_time=df_hotel.groupby("hotel").mean()["days_in_waiting_list"].reset_index()
avg_waiting_time.columns=["Hotel","Avg_Waiting_Time"]
print (avg_waiting_time)

plt.subplot(1,2,1)
sns.barplot(x=avg_waiting_time["Hotel"],y=avg_waiting_time["Avg_Waiting_Time"],palette="tab10")
plt.title("Hotel type vs Average Waiting Time")
plt.ylabel("Average Waiting Time")
plt.xlabel("Hotel type")
plt.show()

* **As city hotels are preferred most by guests , it's having more waiting period.**

# **Customer wise analysis**
*Q1. Find out count of customers on basis of customer type.

In [ ]:
#Q1. Find out count of customers on basis of customer type.
customer=df_hotel["customer_type"].value_counts().reset_index()
customer.columns=["Customer_type","number_of_customers"]
customer

In [ ]:
plt.figure(figsize=(15,6))
plt.subplot(1,2,1)
sns.barplot(x=customer["Customer_type"],y=customer["number_of_customers"],palette="Set2")
plt.title("Customer_type vs NO of customers")
plt.subplot(1,2,2)
plt.pie(df_hotel["customer_type"].value_counts(),autopct="%1.1f%%",shadow=True,radius=1.5,explode=[.2]*4)
plt.legend(labels=df_hotel["customer_type"].value_counts().index.tolist(),loc=(1.1,.65))
plt.show()

* **Here we can see the maximum number of customers are from transient category which is near about 75.1%.**

# **counrty origin customer analysis**
*Q1. Analyse the data from which country guests are visiting most.

In [ ]:
#Q1. Analyse the data from which country guests are visiting most.
country=df_hotel["country"].value_counts().reset_index()[:10]
country.columns=["country_name","number_of_guests"]
country

In [ ]:
plt.pie(country["number_of_guests"],autopct="%1.1f%%",shadow=True,radius=1.3,explode=[.1]*10)
plt.legend(labels=country["country_name"].value_counts().index.tolist(),loc=(1.3,.65))
plt.show()

**Below top 5 countries from which most guests are visiting**
* **1.PORTUGAL-->38.4%**
* **2.GREAT BRITAIN-->14.6%**
* **3.FRANCE---->12.4%**
* **4.ESP--->10.2%**
* **5.DEU---->7.5%**

# **Distributor Channel Analysis**
* Q1. Find out which distribution channel is giving most booking business.
* Q2. Which distribution channel is giving more business to hotels on basis of hotel type?

In [ ]:
#Q1. Find out which distribution channel is giving most booking business.
distribution_channel=df_hotel["distribution_channel"].value_counts().reset_index()
distribution_channel.columns=["DistributionType","NumberOfBookings"]
distribution_channel

In [ ]:
plt.pie(distribution_channel["NumberOfBookings"],autopct="%1.1f%%",shadow=True,radius=1,explode=[.1]*5)
plt.legend(labels=distribution_channel["DistributionType"].value_counts().index.tolist(),loc=(1,.65))
plt.show()

In [ ]:
#Q2. Which distribution channel is giving more business to hotels on basis of hotel type?
dc_hotel = df_hotel.groupby(['distribution_channel', 'hotel'])
df_dc_hotel = pd.DataFrame(round((dc_hotel['adr']).agg(np.mean),2)).reset_index().rename(columns = {'adr': 'avg_adr'})
plt.figure(figsize = (7,5))
sns.barplot(x = df_dc_hotel['distribution_channel'], y = df_dc_hotel['avg_adr'], hue = df_dc_hotel['hotel'])
plt.ylim(40,140)
plt.xlabel('Distribution Channel')
plt.ylabel('Average ADR')
plt.show()

**As a result, it has been found that number of bookings are comming from**
* **1.TA/TO     -->69141 i.e. 79.1%**
* **2.Direct    -->12988 i.e. 14.9%**
* **3.corporate -->5081  i.e. 5.8%**
* **4.GDS       -->181   i.e. 0.2%**
* **5.undefined -->5     i.e. close to 0.001%**

# **Agent wise bookings Analysis**

In [ ]:
#Remove company data where company is not associated with hotel
df_hotel.drop(df_hotel[(df_hotel['agent'] == 0)].index, inplace=True)
#Find count of each agent to predict which agent has given maximum business which can be considered while providing commission to agent.
df_Agent = df_hotel['agent'].value_counts().reset_index().rename(columns = {'index':'Agent','agent':'Agent_wise_bookings'}).sort_values(by = 'Agent_wise_bookings', ascending = False)
df_Agent = df_Agent.nlargest(10, ['Agent']) # Find top 5 agents providing maximum business.
plt.figure(figsize = (10,5))
#print (df_Agent)

sns.barplot(x = 'Agent', y = 'Agent_wise_bookings', data = df_Agent, order = df_Agent.sort_values('Agent_wise_bookings', ascending = False).Agent)

* **Agent ID 531 is giving maximum hotel bookings so this data can be utilized to decide commission % for agent**

# **Company wise bookings Analysis**

In [ ]:
#Remove company data where company is not associated with hotel
df1.drop(df1[(df1['company'] == 0)].index, inplace=True)
#Find company wise hotel count. so if we considered it's a chain of hotel associated with any company then this data will 
#show which company hotels are preferable.
df1['company'].value_counts()
#Find count of each company to predict which company is getting maximum business.
df_Company = df1['company'].value_counts().reset_index().rename(columns = {'index':'Company','company':'Company_Owned_Hotels'}).sort_values(by = 'Company_Owned_Hotels', ascending = False)
df_Company = df_Company.nlargest(10, ['Company_Owned_Hotels']) # Find top 5 agents providing maximum business.
plt.figure(figsize = (10,5))
print (df_Company)
sns.barplot(x = 'Company', y = 'Company_Owned_Hotels', data = df_Company, order = df_Company.sort_values('Company_Owned_Hotels', ascending = False).Company)

* **Above graph shows companies owning maximum no of hotels**

# **Market Segment -Booking Analysis**


In [ ]:
#Find count of each market segment to predict which segment is popular.
df_Market_Segment = df_hotel['market_segment'].value_counts().reset_index().rename(columns = {'index':'MarketSegment','market_segment':'MarketSegment_Wise_Count'}).sort_values(by = 'MarketSegment_Wise_Count', ascending = False)
plt.figure(figsize = (10,5))
print (df_Market_Segment)
#sns.barplot(x = 'MarketSegment', y = 'MarketSegment_Wise_Count', data = df_Market_Segment, order = df_Market_Segment.sort_values('MarketSegment_Wise_Count', ascending = False).MarketSegment)
df_Market_Segment.groupby(['MarketSegment']).sum().plot(kind='pie',y='MarketSegment_Wise_Count',autopct='%1.0f%%')

* **Online TA is most commonly use market segment for booking purpose.**

# ***Hotel booking cancellation on basis of days_in_waiting_list and required_car_parking_spaces***

***Waiting List Vs Cancellations***

In [ ]:
(sns.FacetGrid(df_hotel, hue = 'is_canceled',height = 5, xlim = (0,100)).map(sns.kdeplot, 'days_in_waiting_list', shade = True).add_legend());


* **There are high chances of cancellation when waiting period is high.**

***Car Parking Vs Cancellation***

In [ ]:
(sns.FacetGrid(df_hotel, hue = 'is_canceled',height = 5, xlim = (0,100)).map(sns.kdeplot, 'required_car_parking_spaces', shade = True).add_legend());

* **There is no impact on cancellation due to car parking space**

# ***Analysis of is_repeated_guest column***

In [ ]:
plt.figure(figsize=(8,6))
sns.countplot(data = df_hotel, x = 'is_repeated_guest').set_title('Display count of repeated guests', fontsize = 20)
plt.show()

* **Above data shows that most of the guests are visiting first time.**

***Number of weekdays booked by distribution channel***

In [ ]:
#Number of weekdays booked by distribution channel
plt.figure(figsize=(12,6))
sns.boxplot(x = "distribution_channel", y = "stays_in_week_nights", data = df_hotel, hue = "hotel", palette = 'Set3');
plt.title('No of weekdays Vs Distribution Channel')
plt.ylabel('Stays in week nights')
plt.tight_layout()


* **Visitors from direct and corporate distribution channel are staying almost in same range numbers of week nights.**
* **TA/TO distribution channel has some deviation over stays week-nights between Resort and City hotels.**
* **Undefined and GDS distribution channel visitors had not shown interest in the Resort Hotel.**

***Number of weekend nights booked by distribution channel***

In [ ]:
#Number of weekend nights booked by distribution channel
plt.figure(figsize=(12,6))
sns.boxplot(x = "distribution_channel", y = "stays_in_weekend_nights", data = df_hotel, hue = "hotel", palette = 'Set3');
plt.title('No of weekend night Vs Distribution Channel')
plt.ylabel('Stays in weekends nights')
plt.tight_layout()


* **Direct distribution channel visitors prefer to stay more weekend nights in the Resort Hotel type.**
* **Visitors through Corporate and TA/TO distribution channel are equally preferring between Resort and City hotels.**